doc/develop/api/api_conventions.rst

```diff
@@ -0,0 +1,102 @@
+    located in Zephyr tree. The context where the API is called or implemented is well defined. For
+    example, functions prefixed with `arch_` are intended for use by the Zephyr kernel to invoke
+    architecture specific code. An API is classified as an Internal on a case by case basis.
+    Internal APIs should not be used out-of-tree code. Internal APIs are not subject to
+    ref:`api_lifecycle`. Therefore, can be changed or removed at any time. Changes to the internal
+    APIs must be documented in the migration guide.
+
+- Public
+    These APIs intended for use from any :term:`software component`. Public APIs may be used in-tree
+    and out-of-tree. Public APIs are subject to the :ref:`api_lifecycle`. Therefore, changes to an
+    API are introduced and documented according to the rules defined for the API life state.
+
+Note, that only APIs defined in Zephyr tree are subject to :ref:`api_lifecycle`. External projects
+used as :ref:`modules`may define their own rules for API lifecycle.
+
+Zephyr is a constantly evolving project and API classification may change over time. A Private or
+Internal API may be promoted to Internal or Public API, respectively. Zephyr users are encouraged to
+follow :ref:`rfcs` process to recommend changes in API classification.
+
+The following sections provide guidelines on how to identify the class of an API depending on its
+type.
+
+Header files contents
+=====================
+- Private - functions and data types declared in header files located in `include/zephyr/private`.
+  In addition, private symbols are prefixed with `z_`. Due to historical reasons some APIs prefixed
+  with `z_` are public.
+- Internal - functions and data types declared `include/zephyr/internal`. In addition, Internal APIs
+  must use @internal doxygen command.
+- Public - functions and data types declared in header files located in `include/zephyr`.
+
+In addition, the following prefixes are reserved by Zephyr kernel for use in Zephyr Public APIs:
+
+.. list-table:: Prefixes and Descriptions
+   :header-rows: 1
+   :widths: 10 40 40
+   :stub-columns: 1
+
+   * - Prefix
+     - Description
+     - Example
+   * - atomic_
+     - Denotes an atomic operation.
+     - `atomic_inc`
+   * - device_
+     - Denotes an API relating to devices and their initialization.
+     - `device_get_binding`
+   * - irq_
+     - Denotes an IRQ management operation.
+     - `irq_disable`
+   * - isr_
+     - Denotes an operation called by an Interrupt Service Routine.
+     - `isr_event_send`
+   * - k_
+     - Kernel-specific function.
+     - `k_malloc`
+   * - k_do_
+     - Kernel-specific functions indicating essential operation within the kernel space.
+     - `k_do_event_signal`
+   * - sys_
+     - Catch-all for APIs that do not fit into the other namespaces.
+     - `sys_write32`
+
+Kconfig symbols
+===============
+Kconfig symbols that are promptless and hidden are Internal. All other symbols are Public. The
```

Comment:
that's a really poor signal for the intent of the symbol, it usually just means that the symbol is a proxy for something but these are used for stuff like the i2c bit bang library, the keyboard scanning library, the M_CAN and SJA1000 support libraries... it would not be fair to just trash those symbols just because they have no symbols. This was talked about at length in the arch wg and there' really no easy way out at this point on such a flat namespace.

Reply:
Could you please be more specific what is your suggestion exactly?

Reply:
Some hidden symbol should be considered stable as they are very much usable by the application. I don't really have a suggestion of what should not be considered stable though, the config system does not really help with that.

Reply:
Should the language here just indicate that "Kconfig symbols are Public"?

Reply:
I have no objections to that.

Reply:
Please make the edits here to indicate that all Kconfig symbols are Public.